### ISSUE TITLE
Reference resolver in v2

### ISSUE THREAD

**Issue by derberg** [asyncapi/parser-js#761]
In Parser v1 we use `@apidevtools/json-schema-ref-parser` and now I see that v2 uses `@stoplight/json-ref-resolver` which is marked as deprecated and points to `@apidevtools/json-schema-ref-parser` 😄 

![Screenshot 2023-04-26 at 08 36 52](https://user-images.githubusercontent.com/6995927/234489922-6a1ccf19-35d2-4d34-9bd2-585a4d5b3ef0.png)

any particular reason for that?

cc @magicmatatjahu @smoya

**Comment by smoya** [comment]
TBH, I didn't know it was deprecated. In fact, it seems [we are not the only one](https://github.com/stoplightio/json-ref-resolver/issues/210).

Besides that, there is a reason we use that resolver. That's the ref resolver that Spectral uses and requires when instantiating it. See https://github.com/stoplightio/spectral/blob/66f769001f045eb841f4fa35821a83ce0b661074/packages/core/src/types/spectral.ts#L5-L7.

It is anything but ideal, but we had to use it. In fact, it has cons such as [circular references not being handled](https://github.com/asyncapi/parser-js/issues/599) which have been fixed with a workaround @magicmatatjahu did.

There is an effort by Stoplight team for using `@apidevtools/json-schema-ref-parser` instead. In fact, it seems they started using it but in other tools of their ecosystem. There is an open issue (where @magicmatatjahu also commented time ago)  https://github.com/stoplightio/spectral/issues/1054.

I think the benefit on using Spectral is way far (better) than having to use this deprecated ref resolver. So i will say we should just deal with it and either 🙏 some day Stoplight finally does the migration or either we do it in their code base.

cc @magicmatatjahu Just in case, check my comment if I'm missing something or I said something wrong, as you were the one who took care of that part ❤️

**Comment by smoya** [comment]
> it is not clear why we actually need a custom resolvers to pass to Spectral

Allowing custom resolvers is needed since tools like the generator use it in current implementation. See https://github.com/asyncapi/parser-js/pull/593

**Comment by smoya** [comment]
> no no, don't get me wrong, I love the idea of custom resolvers and this is my main reason why I do not like stoplight resolver and that we do not use the resolver we used in v1 -> [asyncapi/spec#930](https://github.com/asyncapi/spec/issues/930)

We have the option of creating a PR to Spectral with the needed migration for using the `@apidevtools/json-schema-ref-parser`. We could focus on that for the Parser-JS `v2.1.0` maybe?

**Comment by derberg** [comment]
yup, that should be ok. The issue I linked is something we fix after spec 3.0 release, so still few months ahead.

I wanted to make sure we are not going away from the reference resolver.

btw, looking at the custom resolver issue, circular refs issue, the duck taping we had to do. Also the fact that we need to maintain AsyncAPI rules on our own.....does it mean that officially spectral folks like to write https://www.asyncapi.com/blog/creating-consistency-announcing-asyncapi-spectral-together but in fact there is not much help from their side to fix outstanding issues?

**Comment by char0n** [comment]
Yes indeed, very big problem

**Comment by char0n** [comment]
@kennethaasan yep, but you're not able to do that in asyncapi parser-js library (AFAICT) which uses spectral internally.

**Comment by jonaslagoni** [comment]
With spectral there are rules that are applied before and after dereferencing. That is why parser currently only uses spectral, cause the input needs to contain references to correctly apply rules.

**Comment by derberg** [comment]
still a problem
